### DISCUSSION TITLE
Multiple changeable tokens

### DISCUSSION THREAD

**Discussion by Thomas-Bibby** [reach-sh/reach-lang#1077] [Q&A] [answered]
I am trying to have a contract on Algorand where it accepts multiple tokens and these tokens can then be updated in a function. I'm using API structure for the contract.

I have my deployer initialising two tokens and using these to pay works perfectly its just when I go to use tokens that change it starts to mess up. My code is in the zip.
[index.zip](https://github.com/reach-sh/reach-lang/files/8810158/index.zip)

**Comment by jeapostrophe** [comment, marked as answer]
Hi @Thomas-Bibby,

I'm trying to understand what you mean by "changing tokens".

Here's a guess...

I open a store and initially I accept US dollars and Japanese yen. A few years go by and I decide that I will stop accepting yen and start accepting Mexican pesos instead. Even later, I decide to drop dollars and replace them with euros.

Is that what you're imaginging?

If so, then my guess is that you are hitting error RE0103 --- https://docs.reach.sh/rsh/errors/#RE0103 --- because Reach's verification engine cannot, right now, prove the token linearity property if the tokens are not known throughout the program's life. While loops destroy static information, because all information gets compressed into the invariant.

Intuitively, Reach is going to prove that your store has no money left over when it closes. If you switch from dollars to euros, we would need to prove that when you did that, you had zero dollars (and zero euros), but the current verification engine is not sophisticated enough to do this.

Until we figure out how to add this feature, I suggest revising your application to not need this pattern. For example, you could make a pair of applications, one is the storefront and the other is the register. The storefront has a register and the register is in a storefront. When the store wants to start taking different tokens, it replaces the register. The register allows the storefront to take all the money out and put (some) of it in another register.

**Reply by Thomas-Bibby** [reply]
Thanks for the detailed reply. Yes That's exactly what I wanted, thanks for letting me know.

**Reply by jeapostrophe** [reply]
I would be happy to help you on designing the revision I suggest. I'm sorry that you have to adapt in this way to use Reach.

**Reply by Thomas-Bibby** [reply]
All good I have already have a solve. I am having issues with calling a function from another Contract and am having issues, if you could have a quick look at my discord message that would be great https://discord.com/channels/628402598663290882/749639931399241792/983926076638134292

**Reply by jeapostrophe** [reply]
For some reason, the link isn't working for me... would you mind copying in the offending code and error and question into this discussion? It will also be more searchable.

**Reply by Thomas-Bibby** [reply]
I have a the following code on contract A calling contract B: 

```
const result = claimsPoolContract.addFundsInsuranceCover.pay([0, [token1Amt, token1], [0, token2]])(token1Amt * token1Disabled);.
ClaimsPoolContract is declared as const claimsPoolContract = remote(claimsPoolctcInfo, C_interface); 
```

I have also tried 

`const result = claimsPoolContract.addFundsInsuranceCover(amount);`

with C_interface being 
```
const C_interface = {
    roomForInsurance: Fun([UInt], Bool), // View
    addFundsInsuranceCover: Fun([UInt], UInt), // Function
  };.
```

I already can get view data from the claimsPoolContract using roomForInsurance but just can't call addFundsInsurnaceCover. I get an API call failed error in front-end when I run the function.

On Contract B I have a simple function
```
.api(
        User.addFundsInsuranceCover,
        (amt) => {
        },
        (amt) => [0, [amt, defualtToken], [0, lpToken]],
        (amt, k) => {
          const newTotalFunds = totalFunds + amt;
          const newfundsInsured = fundsInsured + amt;

          k(newTotalFunds);
          return [
            claimsPoolThreshold,
            newfundsInsured,
            lpSupply, 
            newTotalFunds
          ];
        }
      )
```

I have tested it ensuring the tokens on both contract are the exact same.

The error I get in front-end
![image](https://user-images.githubusercontent.com/62004466/172618978-e0d1fb6b-a088-465b-b723-a24b17495d1b.png)

**Reply by jeapostrophe** [reply]
This error means you're calling a function that doesn't exist. What is the definition of `User`? I bet the function is actually called `User_addFundsInsuranceCover`
